text/0000-early-payment-proofs.md

```diff
@@ -0,0 +1,164 @@
+
+- Title: early-payment-proofs
+- Authors: [John Tromp](mailto:john.tromp@gmail.com)
+- Start date: Oct 7, 2020
+- RFC PR: Edit if merged: [mimblewimble/grin-rfcs#0000](https://github.com/mimblewimble/grin-rfcs/pull/0000) 
+- Tracking issue: [Edit if merged with link to tracking github issue]
+---
+
+## Summary
+[summary]: #summary
+
+Support generating and validating payment proofs for all transactions, including a timestamp and memo field.
+
+## Motivation
+[motivation]: #motivation
+
+Payment proofs prevent a payment receiver from claiming they didn't receive payment.
+Such dispute/fraud prevention is an essential ingredient to commercial adoption.
+
+Former payment proofs used in Grin didn't apply to invoice flow, at least not
+without additional rounds of communication, which made invoices a difficult proposition.
+
+They also lacked the ability to specify the time and purpose of payment.
+
+This RFC changes the transaction building process so that
+payees commit to a proof of payment promise before the payer signs for the transaction.
+
+## Community-level explanation
+[community-level-explanation]: #community-level-explanation
+
+A payment receiver, in their earliest round of communication to a payment sender,
+shall provide a signature that makes the following promise:
+appearance of certain on-chain data satisfying certain conditions
+shall prove payment for a specified purpose.
+
+This early provision of a payment proof promise enables its use in all possible transaction building flows.
+
+To avoid confusion, the invoice flow shall be called receiver-sender-receiver (RSR) flow and
+the "regular" flow shall be called sender-receiver-sender (SRS) flow.
+
+The first byte of signed data will be used as payment proof type or version (similar to how we encode kernel features).
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Payment Proof is a Witnessed Promise
+
+Since the receiver signature is a promise that payment will be considered proven by 
+certain on-chain data satisfying certain conditions, we shall call such data a witness,
+and let a payment proof consist of a promise paired with a witness.
+
+### Slate changes
+
+To accomodate the various proof types, the slate will include the following related fields:
+
+* `receiver_address` - An ed25519 public key for the receiver, typically the public key of the user's v3 onion address.
+* `timestamp` - The time at which the receiver generates the payment promise
```

Comment:
Not urgent, I'm just considering an experimental implementation of this now to correspond with unified transaction flow. 

`timestamp` should be unambiguously defined here, assuming it's epoch time?

Reply:
Yes, the same 64-bit timestamps appearing in Grin headers.